api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,224 @@
+	MachineDrainRuleDrainBehaviorDrain MachineDrainRuleDrainBehavior = "Drain"
+
+	// MachineDrainRuleDrainBehaviorSkip means the drain for a Pod should be skipped.
+	MachineDrainRuleDrainBehaviorSkip MachineDrainRuleDrainBehavior = "Skip"
+)
+
+// MachineDrainRuleSpec defines the spec of a MachineDrainRule.
+type MachineDrainRuleSpec struct {
+	// drain configures if and how Pods are drained.
+	// +required
+	Drain MachineDrainRuleDrainConfig `json:"drain"`
+
+	// machines defines to which Machines this MachineDrainRule should be applied.
+	//
+	// If machines is not set, the MachineDrainRule applies to all Machines in the Namespace.
+	// If machines contains multiple selectors, the results are ORed.
+	// Within a single Machine selector the results of selector and clusterSelector are ANDed.
+	// Machines will be selected from all Clusters in the Namespace unless otherwise
+	// restricted with the clusterSelector.
+	//
+	// Example: Selects control plane Machines in all Clusters or
+	//          Machines with label "os" == "linux" in Clusters with label
+	//          "stage" == "production".
+	//
+	//  - selector:
+	//      matchExpressions:
+	//      - key: cluster.x-k8s.io/control-plane
+	//        operator: Exists
+	//  - selector:
+	//      matchLabels:
+	//        os: linux
+	//    clusterSelector:
+	//      matchExpressions:
+	//      - key: stage
+	//        operator: In
+	//        values:
+	//        - production
+	//
+	// +optional
+	// +listType=atomic
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=32
+	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))",message="entries in machines must be unique"
+	Machines []MachineDrainRuleMachineSelector `json:"machines,omitempty"`
+
+	// pods defines to which Pods this MachineDrainRule should be applied.
+	//
+	// If pods is not set, the MachineDrainRule applies to all Pods in all Namespaces.
+	// If pods contains multiple selectors, the results are ORed.
+	// Within a single Pod selector the results of selector and namespaceSelector are ANDed.
+	// Pods will be selected from all Namespaces unless otherwise
+	// restricted with the namespaceSelector.
+	//
+	// Example: Selects Pods with label "app" == "logging" in all Namespaces or
+	//          Pods with label "app" == "prometheus" in the "monitoring"
+	//          Namespace.
+	//
+	//  - selector:
+	//      matchExpressions:
+	//      - key: app
+	//        operator: In
+	//        values:
+	//        - logging
+	//  - selector:
+	//      matchLabels:
+	//        app: prometheus
+	//    namespaceSelector:
+	//      matchLabels:
+	//        kubernetes.io/metadata.name: monitoring
+	//
+	// +optional
+	// +listType=atomic
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=32
+	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))",message="entries in pods must be unique"
```

Comment:
one note on these, what's the minimum cluster version that these would work in?

Reply:
1.28. We cannot use CEL, missed that. CI on main would have caught it eventually..

Reply:
Where did you get 1.28 from?

I think this validation would work from 1.26 clusters onwards

Reply:
Sorry. Misread the comment from Vince. v1.28 is the miminum Kubernetes mgmt cluster version that Cluster API v1.9 will support (source: https://main.cluster-api.sigs.k8s.io/reference/versions#core-provider-cluster-api-controller)

According to the Kubernetes website validation rules have become stable with Kubernetes 1.29: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules

Reply:
Not sure what that means for our Kubernetes version support. I think currently we just state that we support >= 1.28.

We didn't specifyi details like. "We support 1.28, but only if you don't disable CEL"

Reply:
Hmm, good point. I would wager most people don't disable beta, enabled by default Kube features, but, we could in theory break someone by adding this.

What release is this going into? Could a release note be added for this release saying we support 1.28 but you must not have disabled CEL?

Reply:
This goes into CAPI v1.9 in December.

Random example:
> Starting with 1.30 kubernetes version and 1.27 LTS versions the beta APIs will be disabled by default when you upgrade to them.
> https://learn.microsoft.com/en-us/azure/aks/upgrade-aks-cluster?tabs=azure-cli#before-you-begin

(but this is not about CEL, it's only for >= 1.30 and I don't know if what they do with beta **APIs** is comparable to beta feature gates)

Reply:
Damn you AKS 😂  Do we risk it? 

Or do we implement the same CEL validations as webhooks for now, and leave the markers there (disable them somehow?) with a todo to reimplement them and delete the code once the minimum kube version is bumped to 1.29 for CAPI?

Reply:
I would suggest to keep it simple and do the following:
* Implement the validation in the webhook for now
* Open a follow-up issue to track the API validation discussion in general and also a specific sub-task for this validation

Context: current plan is that CAPI v1.10 supports mgmt clusters >= v1.29 (which means that once we branched away release-1.9 in 3 weeks this is not a blocker anymore anyway).